### PR TITLE
input: implemented gamepad simulation keyboard and mouse.

### PR DESCRIPTION
**Pull request by js-john**
IMG_0035

### CODE REVIEW

File: Views/VMDisplayMetalViewController.m [hunk 1]
```diff
@@ -199,7 +200,13 @@ - (void)virtualMachine:(UTMVirtualMachine *)vm transitionToState:(UTMVMState)sta
 
 #pragma mark - Helper Functions
 
+- (int) keyValueToScancode: (NSString *) keyValue {
```
Comment (osy86) [inline Views/VMDisplayMetalViewController.m]: What is this for?

File: Views/VMDisplayMetalViewController.h [hunk 2]
```diff
@@ -47,6 +48,23 @@ NS_ASSUME_NONNULL_BEGIN
...
+    BOOL _dpadDownPressed;
+    BOOL _dpadLeftPressed;
+    BOOL _dpadRightPressed;
+    BOOL _buttonAPressed;
+    BOOL _buttonBPressed;
+    BOOL _buttonXPressed;
+    BOOL _buttonYPressed;
+    BOOL _leftShoulderPressed;
+    BOOL _rightShoulderPressed;
+    NSDictionary *_SCANCODE_DICT;
```
Comment (osy86) [inline Views/VMDisplayMetalViewController.h]: Is there a reason a constant dict should be stored as an instance variable?

File: Views/VMDisplayMetalViewController+Gamepad.m [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+        // right trigger
+        if (gamepad.rightTrigger == element && gamepad.rightTrigger.isPressed) {
+            message = @"Right Trigger";
+        }
+        
+        // left shoulder button
+        if (gamepad.leftShoulder == element) {
+            if (gamepad.leftShoulder.isPressed != self->_leftShoulderPressed) {
+                [weakSelf keyboardAndMouseSender:@"leftShoulder" withKeyStatus:gamepad.leftShoulder.isPressed];
+                self->_leftShoulderPressed = gamepad.leftShoulder.isPressed;
```
Comment (osy86) [inline Views/VMDisplayMetalViewController+Gamepad.m]: Shouldn’t all “self” be “weakSelf” inside this block?

**Comment by js-john** [comment]
https://youtu.be/d3n1hb4OKmw

**Comment by js-john** [comment]
I haven’t written iOS code for a long time...😂
I will resubmit it after correcting the mistakes tomorrow...

**Comment by js-john** [comment]
I will resolve the conflicts and re-PR again today.

**Comment by osy86** [comment]
You know that you can `git push --force origin gamepad` without having to open a new PR right? That’s preferred as the comments and code review are preserved.

**Comment by js-john** [comment]
> You know that you can `git push --force origin gamepad` without having to open a new PR right? That’s preferred as the comments and code review are preserved.

fatal: unable to access 'https://github.com/utmapp/UTM.git/': The requested URL returned error: 403
I think I have no access right to push

**Comment by osy86** [comment]
Please update and rebase on master: https://stackoverflow.com/questions/7244321/how-do-i-update-a-github-forked-repository#7244456

Also for the GitHub actions, you need to create a new secret `XCODE_PATH` with the value `/Applications/Xcode_11.4.app`  Thanks!

**Pull request closed without merging**